Sources/App/Core/PackageContributors.swift

```diff
@@ -0,0 +1,132 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import ShellOut
+import Vapor
+
+
+struct PackageAuthors : Encodable, Decodable, Equatable {
+    var authors : [Author]
+    var numberOfContributors : Int
+}
+
+enum PackageContributors {
+
+    /// Extracts the possible authors of the package according to the number of commits.
+    /// A contributor is considered an author when the number of commits is at least a 60 percent
+    /// of the maximum commits done by a contributor.
+    /// - Parameters:
+    ///   - gitCacheDirectoryPath: path to the cache directory where the clone of the package is stored
+    ///   - packageID: the UUID of the package
+    /// - Returns: PackageAuthors
+    static func extract(gitCacheDirectoryPath: String, packageID: UUID?) throws -> PackageAuthors {
+        let contributorsHistory = try GitHistoryLoader.loadContributorsHistory(gitCacheDirectoryPath: gitCacheDirectoryPath, packageID: packageID)
+        let authors = primaryContributors(candidates: contributorsHistory, threshold: 0.6)
+
+        return PackageAuthors(authors: authors.map { Author(name: $0.name) },
+                              numberOfContributors: contributorsHistory.count - authors.count)
+    }
+
+
+    struct Contributor {
+        /// Total number of commits
+        let numberOfCommits: Int
+        let name: String
+    }
+
+    /// Loads the contributors history from a Git repository
+    struct GitHistoryLoader {
+
+        static func loadContributorsHistory(gitCacheDirectoryPath: String, packageID: UUID?) throws -> [Contributor] {
+            do {
+                let commitHistory = try queryGitHistory(gitCacheDirectoryPath: gitCacheDirectoryPath, packageID: packageID)
+                return try parseGitHistory(log: commitHistory)
+            } catch {
+                throw AppError.analysisError(packageID, "loadContributorsHistory failed: \(error.localizedDescription)")
+            }
+        }
+
+        /// Gets the git history in a string log
+        private static func queryGitHistory(gitCacheDirectoryPath: String, packageID: UUID?) throws -> String {
+
+            if !Current.fileManager.fileExists(atPath: gitCacheDirectoryPath) {
+                throw AppError.cacheDirectoryDoesNotExist(packageID, gitCacheDirectoryPath)
+            }
+
+            // attempt to shortlog
+            do {
+                return try Current.git.shortlog(gitCacheDirectoryPath)
+            } catch {
+                throw AppError.shellCommandFailed("gitShortlog",
+                                                  gitCacheDirectoryPath,
+                                                  "queryGitHistory failed: \(error.localizedDescription)")
```

Comment:
Since we map all `Current.git.shortlog` errors to `invalidString` the localised description here won't provide much value. Perhaps it would be worth throwing the original error in the shortlog function so we have a better understanding of why it failed?

Reply:
@Sherlouk could you elaborate a bit more here what you mean :)

Reply:
The `Current.git.shortlog` function has a do/catch in it, which means regardless of the actual error which is thrown, the same 'invalidString' error is caught in the function this review comment is attached to.

You're including the `error.localizedDescription` in the `shellCommandFailed` but this will always be 'invalidString', rather than the actual error from the shell command.

My proposal is that we get rid of the do/catch in the shortlog function itself (not where this comment is). This means the shellCommandFailed error message will contain the actual reason for the error.